pkg/plugins/layout.go

```diff
@@ -0,0 +1,614 @@
+		}
+	}
+
+	return dev.computeFreeSpace(), nil
+}
+
+func (dev Disk) computeFreeSpace() uint {
+	if len(dev.Parts) > 0 {
+		lastPart := dev.Parts[len(dev.Parts)-1]
+		return dev.LastS - (lastPart.StartS + lastPart.SizeS - 1)
+	} else {
+		// Assume first partitions is alined to 1MiB
+		return dev.LastS - (1024*1024/dev.SectorS - 1)
+	}
+}
+
+func (dev Disk) computeFreeSpaceWithoutLast() uint {
+	if len(dev.Parts) > 1 {
+		part := dev.Parts[len(dev.Parts)-2]
+		return dev.LastS - (part.StartS + part.SizeS - 1)
+	} else {
+		// Assume first partitions is alined to 1MiB
+		return dev.LastS - (1024*1024/dev.SectorS - 1)
+	}
+}
+
+//Size is expressed in MiB here
+func (dev Disk) AddPartition(label string, size uint, fileSystem string, pLabel string, console Console) (string, error) {
+	gd := NewGdiskCall(dev.String())
+	pType := "8300"
+	if fatFS, _ := regexp.MatchString("fat|vfat", fileSystem); fatFS {
+		// We are assuming Fat is only used for EFI partitions
+		pType = "EF00"
+	}
+
+	//Check we have loaded partition table data
+	if dev.SectorS == 0 {
+		err := dev.Reload(console)
+		if err != nil {
+			log.Errorf("Failed analyzing disk: %v\n", err)
+			return "", err
+		}
+	}
+
+	var partNum int
+	var startS uint
+	if len(dev.Parts) > 0 {
+		partNum = dev.Parts[len(dev.Parts)-1].Number
+		startS = 0
+	} else {
+		//First partition is aligned at 1MiB
+		startS = 1024 * 1024 / dev.SectorS
+	}
+
+	size = MiBToSectors(size, dev.SectorS)
+	freeS := dev.computeFreeSpace()
+	if size > freeS {
+		return "", errors.New(fmt.Sprintf("Not enough free space in disk. Required: %d sectors; Available %d sectors", size, freeS))
+	}
+
+	partNum++
+	var part = Partition{
+		Number:     partNum,
+		StartS:     startS,
+		SizeS:      size,
+		PLabel:     pLabel,
+		FileSystem: fileSystem,
+		FSLabel:    label,
+		Type:       pType,
+	}
+
+	gd.CreatePartition(&part)
+
+	out, err := gd.WriteChanges(console)
+	if err != nil {
+		return out, err
+	}
+	err = dev.Reload(console)
+	if err != nil {
+		log.Errorf("Failed analyzing disk: %v\n", err)
+		return "", err
+	}
+	err = dev.ReloadPartitionTable(console)
+	if err != nil {
+		log.Errorf("Failed on partprobe: %v\n", err)
+		return "", err
+	}
+	pDev, err := dev.FindPartitionDevice(part.Number, console)
+	if err != nil {
+		return "", err
+	}
+
+	mkfs := MkfsCall{part: part, customOpts: []string{}, dev: pDev}
+	return mkfs.Apply(console)
+}
+
+func (dev Disk) ReloadPartitionTable(console Console) error {
+	out, err := console.Run(fmt.Sprintf("partprobe %s", dev))
+	if err != nil {
+		return errors.New(fmt.Sprintf("Could not reload partition table: %s", out))
+	}
+	return nil
+}
+
+func (dev Disk) FindPartitionDevice(partNum int, console Console) (string, error) {
+	out, err := console.Run(fmt.Sprintf("lsblk -ltnpo name,type %s", dev))
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Could not list device partition nodes: %s", out))
+	}
+
+	re := regexp.MustCompile(fmt.Sprintf("(?m)^(/.*%d) part$", partNum))
```

Comment:
small nit: maybe it's better to use `regexp.Compile` here, otherwise an invalid input could panic the whole execution